### PR TITLE
Flush dbcache early if system is under memory pressure

### PR DESCRIPTION
**Pull request by luke-jr**
No point forcing memory to get pushed out to swap just to cache db changes when we can write the db changes out instead

~~Chasing Concept ACKs and opinions on approach taken~~ (no longer draft)

One eventual goal might be to (on systems where this is supported) allow dbcache to be unlimited.

### CODE REVIEW

File: src/validation.cpp [hunk 1]
```diff
@@ -2313,8 +2313,15 @@ bool CChainState::FlushStateToDisk(
...
         // The cache is large and we're within 10% and 10 MiB of the limit, but we have time now (not in the middle of a block processing).
         bool fCacheLarge = mode == FlushStateMode::PERIODIC && cache_state >= CoinsCacheSizeState::LARGE;
-        // The cache is over the limit, we have to write now.
-        bool fCacheCritical = mode == FlushStateMode::IF_NEEDED && cache_state >= CoinsCacheSizeState::CRITICAL;
+        bool fCacheCritical = false;
+        if (mode == FlushStateMode::IF_NEEDED) {
+            if (cache_state >= CoinsCacheSizeState::CRITICAL) {
+                // The cache is over the limit, we have to write now.
+                fCacheCritical = true;
+            } else if (util::SystemNeedsMemoryReleased()) {
```
Comment (laanwj) [inline src/validation.cpp:2321]: Something I'm slightly worried about is how this is pull, not push. How often does this polling happen? Will it result in a lot of extra syscalls?

**Comment by laanwj** [comment]
Concept ACK I think this is a neat idea.

File: src/validation.cpp (hunk 1 above)
Comment (laanwj) [reply src/validation.cpp:2321]: Please log a message when this happens (possibly with a only-once check against flooding).

**Comment by jb55** [comment]
Concept ACK

File: src/validation.cpp (hunk 1 above)
Comment (luke-jr) [reply src/validation.cpp:2321]: Should be at most once per block.

**Comment by luke-jr** [comment]
Added logging, a config option, and fixed test (and took out of Draft)

**Comment by achow101** [comment]
Closing this as it has not had any activity in a while. If you are interested in continuing work on this, please leave a comment so that it can be reopened.

**Comment by luke-jr** [comment]
This continues to be updated and maintained for Knots.

Feel free to reopen it if anyone shows interest in reviewing. (Ping me if I need to push an older commit temporarily to reopen)

**Comment by andrewtoth** [comment]
Hi @luke-jr I think this is still worth pursuing. Are you still interested in working on this?

**Comment by luke-jr** [comment]
This version will no longer work as expected. Happy to discuss the latest challenges further if you want to take over it, but it is not a primary focus of mine.

**Comment by andrewtoth** [comment]
Sure, I will take a look and see how it can be applied to latest master. Any advice would be appreciated :pray:.

**Pull request closed without merging**